src/shogun/machine/Pipeline.cpp

```diff
@@ -0,0 +1,176 @@
+		REQUIRE(
+		    m_stages.empty() ||
+		        holds_alternative<CTransformer*>(m_stages.back().second),
+		    "Transformers can not be placed after machines. Last element is "
+		    "%s\n",
+		    m_stages.back().first);
+
+		SG_REF(transformer);
+		m_stages.emplace_back(name, transformer);
+
+		return this;
+	}
+
+	CPipeline* CPipeline::then(CMachine* machine)
+	{
+		return then(machine->get_name(), machine);
+	}
+
+	CPipeline* CPipeline::then(const std::string& name, CMachine* machine)
+	{
+		REQUIRE(
+		    m_stages.empty() ||
+		        holds_alternative<CTransformer*>(m_stages.back().second),
+		    "Multiple machines are added to pipeline. Last element is %s\n",
+		    m_stages.back().first);
+
+		SG_REF(machine);
+		m_stages.emplace_back(name, machine);
+
+		return this;
+	}
+
+	bool CPipeline::train_machine(CFeatures* data)
+	{
+		check_pipeline();
+
+		if (train_require_labels())
+		{
+			REQUIRE(m_labels, "No labels given.\n");
+		}
+
+		for (auto&& stage : m_stages)
+		{
+			if (holds_alternative<CTransformer*>(stage.second))
+			{
+				auto transformer = shogun::get<CTransformer*>(stage.second);
+				transformer->train_require_labels()
+				    ? transformer->fit(data, m_labels)
+				    : transformer->fit(data);
+
+				data = transformer->transform(data);
+			}
+			else
+			{
+				auto machine = shogun::get<CMachine*>(stage.second);
+				if (machine->train_require_labels())
+					machine->set_labels(m_labels);
+				machine->train(data);
+			}
+		}
+		return true;
+	}
+
+	CLabels* CPipeline::apply(CFeatures* data)
+	{
+		check_pipeline();
+
+		for (auto&& stage : m_stages)
+		{
+			if (holds_alternative<CTransformer*>(stage.second))
+			{
+				auto transformer = shogun::get<CTransformer*>(stage.second);
+				data = transformer->transform(data);
+			}
+			else
+			{
+				auto machine = shogun::get<CMachine*>(stage.second);
+				return machine->apply(data);
+			}
+		}
+
+		return nullptr; // unreachable
+	}
+
+	bool CPipeline::train_require_labels() const
+	{
+		bool require_labels = false;
+
+		for (auto&& stage : m_stages)
+		{
+			visit(
+			    [&require_labels](auto&& fittable) {
+				    require_labels = fittable->train_require_labels();
+				},
+			    stage.second);
+
+			if (require_labels)
+				return require_labels;
+		}
+
+		return require_labels;
+	}
+
+	std::string CPipeline::to_string() const
+	{
+		std::stringstream ss;
+
+		for (auto i : range(m_stages.size()))
+		{
+			ss << '[' << i << "]: " << m_stages[i].first << std::endl;
+		}
+
+		return ss.str();
+	}
+
+	CTransformer* CPipeline::get_transformer(const std::string& name) const
+	{
+		for (auto&& stage : m_stages)
+		{
+			if (stage.first == name &&
+			    holds_alternative<CTransformer*>(stage.second))
+				return shogun::get<CTransformer*>(stage.second);
+		}
+
+		SG_ERROR("Transformer with name %s not found.\n", name.c_str());
+
+		return nullptr;
+	}
+
+	void CPipeline::check_pipeline()
+	{
+		REQUIRE(!m_stages.empty(), "Pipeline is empty");
+		REQUIRE(
```

Comment:
i wonder if we should start using typed exceptions, because for now EVERYTHING is ShogunException. things like `MachineNotTrainedException` etc....
@lisitsyn @karlnapf mm?

Reply:
Actually, this was part of the "user experience" GSoC project. So yes makes total sense imo